File: backend/src/services/enrollments.service.ts
import { EnrollmentsRepository } from '../repositories/enrollments.repository';

export class EnrollmentsService {
  private enrollmentsRepo: EnrollmentsRepository;

  constructor() {
    this.enrollmentsRepo = new EnrollmentsRepository();
  }

  /**
   * Obtener inscripciones por estudiante
   */
  async getEnrollmentsByStudent(studentId: string): Promise<any[]> {
    return this.enrollmentsRepo.findByStudent(studentId);
  }

  /**
   * Obtener inscripciones por curso
   */
  async getEnrollmentsByCourse(courseId: string): Promise<any[]> {
    return this.enrollmentsRepo.findByCourse(courseId);
  }

  /**
   * Crear inscripción
   */
  async enrollStudent(data: { student_id: string; course_id: string; progress?: number }): Promise<string> {
    // Verificar si ya está inscrito
    const isEnrolled = await this.enrollmentsRepo.isEnrolled(data.student_id, data.course_id);
    if (isEnrolled) {
      throw new Error('Student is already enrolled in this course');
    }

    return this.enrollmentsRepo.createEnrollment({
      student_id: data.student_id,
      course_id: data.course_id,
      progress: data.progress || 0,
      status: 'active'
    });
  }

  /**
   * Eliminar inscripción
   */
  async unenrollStudent(enrollmentId: string): Promise<void> {
    await this.enrollmentsRepo.delete(enrollmentId);
  }

  /**
   * Actualizar progreso de inscripción
   */
  async updateEnrollmentProgress(enrollmentId: string, progress: number): Promise<void> {
    await this.enrollmentsRepo.update(enrollmentId, { progress });
  }
}

File: backend/app/services/enrollments_service.py
"""Enrollments service."""

from app.repositories.enrollments_repository import EnrollmentsRepository


class EnrollmentsService:
    def __init__(self, repository: EnrollmentsRepository | None = None) -> None:
        self._repository = repository or EnrollmentsRepository()

    def list_by_student(self, student_id: str) -> list[dict]:
        return self._repository.list_by_student(student_id)

    def list_by_course(self, course_id: str) -> list[dict]:
        return self._repository.list_by_course(course_id)

    def enroll(self, student_id: str, course_id: str, progress: int = 0) -> str:
        if self._repository.exists(student_id, course_id):
            raise ValueError("Student is already enrolled in this course")
        return self._repository.create(student_id, course_id, progress)

    def unenroll(self, enrollment_id: str) -> None:
        self._repository.delete(enrollment_id)
